crates/sui-framework/sources/balance.move

```diff
@@ -105,16 +109,18 @@ module sui::balance {
     }
 
     /// CAUTION: this function creates a `Balance` without increasing the supply.
-    /// It should only be called by `sui_system::advance_epoch` to create staking rewards,
+    /// It should only be called by the epoch change system txn to create staking rewards,
     /// and nowhere else.
-    public(friend) fun create_staking_rewards<T>(value: u64): Balance<T> {
+    fun create_staking_rewards<T>(value: u64, ctx: &TxContext): Balance<T> {
+        assert!(tx_context::sender(ctx) == @0x0, ENotSystemAddress);
         Balance { value }
     }
 
     /// CAUTION: this function destroys a `Balance` without decreasing the supply.
-    /// It should only be called by `sui_system::advance_epoch` to destroy storage rebates,
+    /// It should only be called by the epoch change system txn to destroy storage rebates,
     /// and nowhere else.
-    public(friend) fun destroy_storage_rebates<T>(self: Balance<T>) {
+    fun destroy_storage_rebates<T>(self: Balance<T>, ctx: &TxContext) {
```

Comment:
Curious, why does this need to take `T` as type param?

Reply:
It doesn't need to since it's only used for T = SUI. Alternatively we can make this public(friend) and balance friend with Sui. And have a function in Sui that does destroy_storage_rebates<SUI>. Either way we need to have a generic fun here that does this. Which way do you think is better?

Reply:
Because in Move you can't have a circular dependency :/